pallas-network/src/miniprotocols/localtxsubmission/codec.rs

```diff
@@ -101,24 +104,77 @@ impl Encode<()> for EraTx {
 
 impl<'b> Decode<'b, ()> for RejectReason {
     fn decode(d: &mut Decoder<'b>, _ctx: &mut ()) -> Result<Self, decode::Error> {
-        let remainder = d.input().to_vec();
-        Ok(RejectReason(remainder))
+        d.array()?;
+        d.array()?;
+        let era = d.u8()?;
+        let errors = d.decode()?;
+
+        Ok(RejectReason::EraErrors(era, errors))
     }
 }
 
+// FIXME: Just the scaffold for encoding
 impl Encode<()> for RejectReason {
     fn encode<W: encode::Write>(
         &self,
         e: &mut Encoder<W>,
         _ctx: &mut (),
     ) -> Result<(), encode::Error<W::Error>> {
-        e.writer_mut()
-            .write_all(&self.0)
-            .map_err(encode::Error::write)?;
+        e.array(0)?;
         Ok(())
     }
 }
 
+impl<'b> Decode<'b, ()> for TxError {
+    fn decode(d: &mut Decoder<'b>, _ctx: &mut ()) -> Result<Self, decode::Error> {
+        match d.datatype()? {
+            CborType::U8 => return Ok(TxError::U8(d.u8()?)),
+            CborType::Array => d.array()?,
+            _ => {
+                return Err(decode::Error::message(
+                    "Error types are supposed to be `array` or `u8`",
+                ))
+            }
+        };
+        // This seems to be always 1 for Conway errors.
+        let fst_num = d.u8()?;
+        let data = d.input();
+        let size = d.array()?.ok_or(decode::Error::message(
+            "Errors are supposed to be definite-length arrays",
+        ))?;
+        let mut raw_elements = Vec::new();
+        let start_pos = d.position();
+
+        if fst_num != 1 {
+            for _ in 0..size {
+                d.skip()?
+            } // Skip decoding the elements of the array
+              // Store the slice of raw CBOR
+            raw_elements.append(&mut data[start_pos..d.position()].to_vec());
+
+            return Ok(TxError::Raw(raw_elements));
+        };
+
+        match d.u8()? {
+            9 => {
+                d.tag()?;
+                let vec_bytes: Vec<Bytes> = d.decode()?;
+
+                Ok(TxError::ExtraneousScriptWitnessesUTXOW(vec_bytes))
+            }
+            _ => {
+                for _ in 1..size {
+                    d.skip()?
+                } // Skip decoding the elements of the array
+                  // Store the slice of raw CBOR
+                raw_elements.append(&mut data[start_pos..d.position()].to_vec());
+
+                Ok(TxError::Raw(raw_elements))
+            }
+        }
```

Comment:
this seems a little bit more "imperative" than what I was hoping for. Instead embedding the decoding logic as conditional branches, let the decoder do it's thing by defining intermediate structures that handle smaller chunks of the logic.

Reply:
Thank you for the insight; I added a missing layer of the error structure. Hope it looks better now!